src/cls/rbd/cls_rbd_types.h

```diff
@@ -411,6 +412,53 @@ struct TrashSnapshotNamespace {
   }
 };
 
+struct MirrorSnapshotNamespace {
+  static const SnapshotNamespaceType SNAPSHOT_NAMESPACE_TYPE =
+    SNAPSHOT_NAMESPACE_TYPE_MIRROR;
+
+  string mirror_local_uuid;
```

Comment:
Looking at this I almost wonder if it will be clearer to just re-use the journaling `mirror_uuid` concept to denote the owner and re-use the magic values to denote demotion. To me it's odd to record your local mirror uuid here even in the case where you aren't the primary (i.e. the primary and non-primary sites will have differences).

Reply:
Is your idea that you would want the corresponding snapshots on primary and non-primary site look the same? I.e. `mirror_uuid` would contain the `uuid` of the primary site and would use for recognizing if it is primary or non-primary? What about `mirror_peers` on the secondary?

Because my idea was that primary and non-primary snapshots are different. Primary snapshots are created by users (or some "automatic snapshoting" mechanism). The users are not allowed to create non-primary snapshots (except the demote request). Those are created by rbd-mirror when copying data. At the moment when a secondary snapshot is created the rbd-mirror knows the snap_id of the remote (primary) snapshot, and could record it in the namespace to store the reference. We could use `tid` instead of `remote_snap_id` for this as you suggested, but then I think we need an additional cls method to generate tids when creating primary snapshots, and I don't see much benefits. `remote_snap_id` just would work for me the same, but if you still prefer `tid` I will not object. 

 `mirror_local_uuid` would be always the uuid of the site where the snapshot is created. I don't see much use of it except the case when mirroring is disabled/enabled for a pool and we want to recognize (and probably get rid off) the stale snapshots. 

Also, the rbd-mirror would know the remote site uuid and could just reuse `mirror_peers` field to store it. So `mirror_peers`  would have different meaning for primary and non-primary snapshots.

A special case would be non-primary snapshots created by demote request. In this case `remote_snap_id` and `mirror_peers` would be unused.

And as I said previously, I would also like to store the snapshot state ("copying started", "copying complete") in the secondary snapshot namespace. But if you prefer to store this state in the image metadata, I am ok with this too. Because we will probably want to store not only the state but also how much objects copied so far, in case we need to restart sync. Not sure if namespace is good for this.

Reply:
> Is your idea that you would want the corresponding snapshots on primary and non-primary site look the same? I.e. `mirror_uuid` would contain the `uuid` of the primary site and would be used for recognizing if it is primary or non-primary? What about `mirror_peers` on the secondary?

Yes, that was the idea since (1) there is no need to store the local mirror uuid within the snapshot if it's always going to be the local mirror uuid and (2) you are going to have to add some details to link back to the true owner of the snapshot anyway. I am fine w/ the `mirror_peers` being different since that isn't used for ownership tracking.

> Because my idea was that primary and non-primary snapshots are different. Primary snapshots are created by users (or some "automatic snapshoting" mechanism). The users are not allowed to create non-primary snapshots (except the demote request). Those are created by rbd-mirror when copying data. At the moment when a secondary snapshot is created the rbd-mirror knows the snap_id of the remote (primary) snapshot, and could record it in the namespace to store the reference. 

Is that also true if we eventually support multiple peers? If peer A is the primary to peer B and C, could peer C ever use the snapshot data in peer B to resync itself?

> We could use `tid` instead of `remote_snap_id` for this as you suggested, but then I think we need an additional cls method to generate tids when creating primary snapshots, and I don't see much benefits. `remote_snap_id` just would work for me the same, but if you still prefer `tid` I will not object.

Why would you need a cls method for the tid? I guess this new form of mirroring isn't tied to exclusive-lock so in theory you could have two concurrent snapshot creators? 

> `mirror_local_uuid` would be always the uuid of the site where the snapshot is created. I don't see much use of it except the case when mirroring is disabled/enabled for a pool and we want to recognize (and probably get rid off) the stale snapshots.

Oh, perhaps I was just confused by the name then.  `mirror_local_uuid` is really just the `mirror_uuid` of the primary? If that's the case, I'm fine w/ it as long as you can properly handle promotion/demotion. 

> Also, the rbd-mirror would know the remote site uuid and could just reuse `mirror_peers` field to store it. So `mirror_peers` would have different meaning for primary and non-primary snapshots.

Ack -- I agree that `mirror_peers` would be different between sites.

> A special case would be non-primary snapshots created by demote request. In this case `remote_snap_id` and `mirror_peers` would be unused.

You would still need to populate the peers for a demote, though, correct? Otherwise, how do you track its propagation between sites so that you know when its safe to delete?

> And as I said previously, I would also like to store the snapshot state ("copying started", "copying complete") in the secondary snapshot namespace. But if you prefer to store this state in the image metadata, I am ok with this too. Because we will probably want to store not only the state but also how much objects copied so far, in case we need to restart sync. Not sure if namespace is good for this.

I guess whatever is cleaner and easier. I had originally suggested just re-using the image-meta for all the mutable stuff trying to avoid new cls methods. Now we have a cls method for unlinking peers from the namespace, I guess another method to update the transient progress status wouldn't be the worst.

Reply:
> Is that also true if we eventually support multiple peers? If peer A is the primary to peer B and C, could peer C ever use the snapshot data in peer B to resync itself?

I think yes. The peer C will look for the remote snap_id and the remote uuid in peer's B  secondary snapshots and for those that match ones it has it will know they are the same sync points. Or do you mean something different? Basically, I can't see how a generated tid would differ from a snap_id, unless it is global across all sites, but I suppose it is not what you mean?

> Why would you need a cls method for the tid? I guess this new form of mirroring isn't tied to exclusive-lock so in theory you could have two concurrent snapshot creators?

I am not sure I understand. We want a monotonously increasing id, shared between clients. I suppose  it is only if it is generated/increased server side? If you have two clients creating mirroring snapshots for the same image (for some reason),  how you can make sure they do not race for new tid?

> Oh, perhaps I was just confused by the name then. mirror_local_uuid is really just the mirror_uuid of the primary?

No, I really meant the uuid of the site where the snapshot is created. But now I seem to agree with your idea of using `mirror_uuid` field instead. I will try this.